cirq-core/cirq/ops/gateset.py

```diff
@@ -0,0 +1,304 @@
+            >>> assert (cirq.X in gate_family) == True
+            >>> assert (cirq.X ** sympy.Symbol("theta") not in gate_family) == False
+
+        b) Type Family:
+            >>> gate_family = cirq.GateFamily(cirq.XPowGate)
+            >>> assert (cirq.X in gate_family) == True
+            >>> assert (cirq.X ** sympy.Symbol("theta") in gate_family) == True
+
+    In order to create gate families with constraints on parameters of a gate
+    type, users should derive from the `cirq.GateFamily` class and override the
+    `predicate` method used to check for gate containment.
+    """
+
+    def __init__(
+        self,
+        gate: Union[Type[raw_types.Gate], raw_types.Gate],
+        *,
+        name: Optional[str] = None,
+        description: Optional[str] = None,
+    ) -> None:
+        """Init GateFamily
+        Args:
+            gate: A python `type` inheriting from `cirq.Gate` for type based membership checks, or
+                a non-parameterized instance of a `cirq.Gate` for equality based membership checks.
+            name: The name of the gate family.
+            description: Human readable description of the gate family.
+
+        Raises:
+            ValueError: if `gate` is not a `cirq.Gate` instance or subclass.
+            ValueError: if `gate` is a parameterized instance of `cirq.Gate`.
+        """
+        if not (
+            isinstance(gate, raw_types.Gate)
+            or (isinstance(gate, type) and issubclass(gate, raw_types.Gate))
+        ):
+            raise ValueError(f'Gate {gate} must be an instance or subclass of `cirq.Gate`.')
+        if isinstance(gate, raw_types.Gate) and protocols.is_parameterized(gate):
+            raise ValueError(f'Gate {gate} must be a non-parameterized instance of `cirq.Gate`.')
+
+        self._gate = gate
+        self._name = name if name else self._default_name()
+        self._description = description if description else self._default_description()
+
+    def _gate_str(self, gettr: Callable[[Any], str] = str) -> str:
+        return (
+            gettr(self.gate)
+            if isinstance(self.gate, raw_types.Gate)
+            else f'{self.gate.__module__}.{self.gate.__name__}'
+        )
+
+    def _default_name(self) -> str:
+        family_type = 'Instance' if isinstance(self.gate, raw_types.Gate) else 'Type'
+        return f'{family_type} GateFamily: {self._gate_str()}'
+
+    def _default_description(self) -> str:
+        check_type = r'g == {}' if isinstance(self.gate, raw_types.Gate) else r'isinstance(g, {})'
+        return f'Accepts `cirq.Gate` instances `g` s.t. `{check_type.format(self._gate_str())}`'
+
+    @property
+    def gate(self) -> Union[Type[raw_types.Gate], raw_types.Gate]:
+        return self._gate
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def description(self) -> str:
+        return self._description
+
+    def predicate(self, gate: raw_types.Gate) -> bool:
```

Comment:
is there a more obvious name we could use here? `belongs()`?

Reply:
I expect most users to use the `g in gate_family` construct for checking containment. I think the name `predicate` might be more instructive for developers overriding the default predicate by implementing new gate families inheriting from this class. We could also make this method private and have a `def _predicate()`.  

I don't have strong opinions here. LMK your thoughts and I'll be happy to change.

Reply:
Made the predicate a protected member to avoid exposing confusing APIs. Now the users should always check for containment via `g in gate_family`.